### PR TITLE
MINOR: move tiered storage related configs to a separate class within LogConfig

### PR DESCRIPTION
**Pull request by ccding**
The original code uses a `RemoteLogManagerConfig` class to store KIP-405 configs and adds three configs to `LogConfig`. This makes the code complicated and developers may be confused.

This PR allows us to access `RemoteLogManagerConfig` from `KafkaConfig` and do the same for `LogConfig`. Kafka developers will see the same interface for the KIP-405 configs. After this change, if we want to read `remoteStorageEnable` we should use `LogConfig.tieredLogConfig.remoteStorageEnable` instead of `LogConfig.remoteStorageEnable`. The same for `localRetentionMs` and `localRetentionBytes`. If we want to read configs in `RemoteLogManagerConfig`, we should use `KafkaConfig.tieredKafkaConfig.xxx`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR. We should include the motivation when suggesting a change.

**Comment by ccding** [comment]
Included the motivation and updated the PR. PTAL @ijuma @junrao @kowshik @satishd

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 1]
```diff
@@ -1434,6 +1434,9 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
   val zkEnableSecureAcls: Boolean = getBoolean(KafkaConfig.ZkEnableSecureAclsProp)
   val zkMaxInFlightRequests: Int = getInt(KafkaConfig.ZkMaxInFlightRequestsProp)
 
+  private val _tieredKafkaConfig = new RemoteLogManagerConfig(this)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Why is this prefixed with `tiered` while the class name is prefixed with `Remote`?

Comment (ccding) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: I had a hard time naming this variable. Would `_remoteLogManagerConfig` be better?

File: core/src/main/scala/kafka/log/LogConfig.scala [hunk 2]
```diff
@@ -107,46 +107,52 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
   val LeaderReplicationThrottledReplicas = getList(LogConfig.LeaderReplicationThrottledReplicasProp)
   val FollowerReplicationThrottledReplicas = getList(LogConfig.FollowerReplicationThrottledReplicasProp)
   val messageDownConversionEnable = getBoolean(LogConfig.MessageDownConversionEnableProp)
-  val remoteStorageEnable = getBoolean(LogConfig.RemoteLogStorageEnableProp)
 
-  val localRetentionMs: Long = {
-    val localLogRetentionMs = getLong(LogConfig.LocalLogRetentionMsProp)
+  class TieredLogConfig {
```
Comment (junrao) [inline core/src/main/scala/kafka/log/LogConfig.scala]: This probably should be named RemoteLogConfig to match RemoteLogManagerConfig?

**Review by junrao — COMMENTED** [review]
@ccding : Thanks for the PR. Just one comment below.

File: core/src/main/scala/kafka/log/LogConfig.scala (hunk 2 above)
Comment (ccding) [reply core/src/main/scala/kafka/log/LogConfig.scala]: Done. Thanks for the comment

File: core/src/main/scala/kafka/log/LogConfig.scala [hunk 3]
```diff
@@ -107,46 +107,52 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
...
 
-    // -2 indicates to derive value from retentionSize property.
-    if(localLogRetentionBytes == -2) retentionSize;
-    else {
-      // Added validation here to check the effective value should not be more than RetentionBytes.
-      if(localLogRetentionBytes == -1 && retentionSize != -1) {
-        throw new ConfigException(LogConfig.LocalLogRetentionBytesProp, localLogRetentionBytes, s"Value must not be -1 as ${LogConfig.RetentionBytesProp} value is set as $retentionSize.")
-      }
+      // -2 indicates to derive value from retentionSize property.
+      if(localLogRetentionBytes == -2) retentionSize;
```
Comment (kowshik) [inline core/src/main/scala/kafka/log/LogConfig.scala]: nit: it seems like we could remove the semicolon at the end.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java [hunk 4]
```diff
@@ -253,9 +253,9 @@ public RemoteLogManagerConfig(AbstractConfig config) {
              config.getInt(REMOTE_LOG_READER_THREADS_PROP),
              config.getInt(REMOTE_LOG_READER_MAX_PENDING_TASKS_PROP),
              config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP),
-             config.originalsWithPrefix(config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP)),
+             config.originalsWithPrefix(config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP) == null ? "" : config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP)),
```
Comment (kowshik) [inline storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: Hmm, why is this change needed? It doesn't seem like the PR is altering behavior such as these but maybe I'm missing something.

**Review by kowshik — COMMENTED** [review]
@ccding Thanks for the PR! LGTM. Just few comments below.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java (hunk 4 above)
Comment (satishd) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: What is the intent for this change here?

**Review by satishd — COMMENTED** [review]
thanks @ccding for the PR. left a comment similar to what Kowshik had.

**Comment by ijuma** [comment]
Thanks for explaining the motivation @ccding. In my opinion, this is a bit confusing. What makes remote log configs special when compared to local log configs? The same arguments regarding reuse, etc. could be said for those, right?

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java (hunk 4 above)
Comment (ccding) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: We will get a null pointer exception when we call `new KafkaConfig()` without setting `REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP`. E.g., https://github.com/apache/kafka/blob/d6f6edd2b17cbaccd4c87bfa66f871a676760044/core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala#L1057

This happens everywhere in the codebase after this change: 
https://github.com/apache/kafka/pull/11110/files#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R1437-R1438

Since `REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP` is not a required configuration if we don't enable KIP-405, I think we should pass an empty string to RemoteLogManagerConfig if it is null.

**Review by junrao — COMMENTED** [review]
@ccding : Thanks for the PR. It looks reasonable to me. Since remote storage is a pluggable module, perhaps putting all remote log related configs under a different namespace reduces the surface area to developers not touching remote logs and makes those configs less likely to be confused with other existing configs. @satishd : What do you think of the overall approach? Thanks.

File: storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java (hunk 4 above)
Comment (satishd) [reply storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java]: @ccding It should be non empty based on the validator that is set for this config. If an empty string is passed then all the Kafka config properties will be passed, which is wrong. 
There should be null validation instead of passing an empty string as mentioned below.  Same for the `REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX_PROP` check. 

```
config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP) != null 
                     ? config.originalsWithPrefix(config.getString(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP)) 
                     : Collections.emptyMap()
```

**Comment by satishd** [comment]
@junrao The intention of the PR  looks reasonable to me. I left a comment [here](https://github.com/apache/kafka/pull/11110#discussion_r696454031) on one of the changes.

**Comment by ccding** [comment]
Thank you @satishd

Addressed your comments.

**Review by satishd — APPROVED** [review]
Thanks @ccding for addressing the review comments. LGTM.

**Review by kowshik — APPROVED** [review]
@ccding: LGTM, thanks for the updated PR.

**Review by junrao — APPROVED** [review]
@ccding : Thanks for the updated PR. LGTM
